types/externs.d.ts

```diff
@@ -51,30 +27,13 @@ declare global {
   };
 
   /** Make optional all properties on T and any properties on object properties of T. */
-  type RecursivePartial<T> = {
-    [P in keyof T]+?:
-      // RE: First two conditions.
-      // If type is a union, map each individual component and transform the resultant tuple back into a union.
-      // Only up to 4 components of a union is supported (all but the last few are dropped). For more, modify the second condition
-      // and `IntersectionOfFunctionsToType`.
-      // Ex: `{passes: PassJson[] | null}` - T[P] doesn't exactly match the array-recursing condition, so without these first couple
-      // conditions, it would fall through to the last condition (would just return T[P]).
-
-      // RE: First condition.
-      // Guard against large string unions, which would be unreasonable to support (much more than 4 components is common).
-
-      SplitType<T[P]> extends string[] ? T[P] :
-      GetLength<SplitType<T[P]>> extends 2|3|4 ? RecursivePartialUnion<T[P]>[number] :
-
-      // Recurse into arrays.
-      T[P] extends (infer U)[] ? RecursivePartial<U>[] :
-
-      // Recurse into objects.
-      T[P] extends (object|undefined) ? RecursivePartial<T[P]> :
-
-      // Strings, numbers, etc. (terminal types) end here.
-      T[P];
-  };
+  type RecursivePartial<T> =
+    // Recurse into arrays and tuples: elements aren't (newly) optional, but any properties they have are.
```

Comment:
this does treat tuples the same as arrays, where sometimes you really might always want an array of a certain length (so you use a tuple for the type) and it's ok if it's `['Document', undefined, undefined, 'Script']` in test code. That seemed unlikely enough that I didn't do it here to keep things simple, but something like the following would catch most of those cases:

```ts
type RecursivePartialArrayOrTuple<T> = T extends (infer U)[] ?
  number extends T['length'] ?
    // Array or unbounded-length tuple.
    RecursivePartial<U>[] :
    // Bounded-length tuple.
    {[P in keyof T]?: RecursivePartial<T[P]>} :
  never;
```
That does still treat unbounded-length tuples (e.g. `[string, boolean, ...number[]]`) like arrays and not tuples, but I'm not sure if there is a way to tell the difference in a conditional type.